### ISSUE TITLE
Cannot use context-less errors with generic source

### ISSUE THREAD

**Issue by TheAifam5** [shepmaster/snafu#225]
Hey,
i am trying to pass a factory function as argument which returns `Result`.

```rust
pub fn initialize<T, F, E>(factory: F) -> Result<(), BootloaderError<E>>
where
  T: Log + 'static,
  F: FnOnce() -> Result<T, E>,
  E: std::error::Error + 'static,
{
  log::set_boxed_logger(Box::new(factory().context(Internal {})?))?;
  log::set_max_level(LevelFilter::Info);

  Ok(())
}
```

### Without Context:

```rust
#[derive(Debug, Snafu)]
pub enum BootloaderError<E>
where
  E: std::error::Error + 'static,
{
  #[snafu(context(false))]
  SetLogger { source: SetLoggerError, backtrace: Backtrace },

  #[snafu(context(false))]
  Internal { source: E },
}
```

Error:

```
error[E0119]: conflicting implementations of trait `std::convert::From<log::SetLoggerError>` for type `bootloader::BootloaderError<log::SetLoggerError>`:
 --> xx/src/bootloader.rs:4:17
  |
4 | #[derive(Debug, Snafu)]
  |                 ^^^^^
  |                 |
  |                 first implementation here
  |                 conflicting implementation for `bootloader::BootloaderError<log::SetLoggerError>`
```

## With context

```rust
#[derive(Debug, Snafu)]
pub enum BootloaderError<E>
where
  E: std::error::Error + 'static,
{
  #[snafu(context(false))]
  SetLogger { source: SetLoggerError, backtrace: Backtrace },

  #[snafu(display("test"))]
  Internal { source: E },
}
```

It just works, but I would like to pass that error further to the caller rather giving additional message.

Best regards,
TheAifam5

**Comment by shepmaster** [comment]
This isn't a SNAFU problem, it's a limitation imposed by Rust. See [How is there a conflicting implementation of `From` when using a generic type?](https://stackoverflow.com/q/37347311/155423).

TL;DR: which code path in the first example should be chosen when `SetLoggerError` **also** implements `std::error::Error + 'static`? Note that Rust doesn't care if it *doesn't* implement it currently.

Please reopen if you think that this doesn't explain the problem.
